pymc_marketing/mmm/base.py

```diff
@@ -78,132 +67,241 @@ def methods(self) -> List[Any]:
+
+        This method applies preprocessing methods to the data ('X' or 'y'), which are specified in the preprocessing_methods property of this object.
+        It iteratively applies each method in the appropriate list (either for 'X' or 'y') to the data.
+
+        Parameters
+        ----------
+        target : str
+            Indicates whether the data represents features ('X') or the target variable ('y').
 
-    @abstractmethod
-    def build_model(self, *args, **kwargs) -> None:
-        raise NotImplementedError()
+        data : pd.DataFrame
+            The data to be preprocessed.
 
-    def get_prior_predictive_data(self, *args, **kwargs) -> az.InferenceData:
+        Returns
+        -------
+        pd.DataFrame
+            The preprocessed data.
+
+        Raises
+        ------
+        ValueError
+            If the target is neither 'X' nor 'y'.
+
+        Example
+        -------
+        >>> data = pd.DataFrame({"x1": [1, 2, 3], "y": [4, 5, 6]})
+        >>> self.preprocess('X', data)
+        """
+        if target == "X":
+            for method in self.preprocessing_methods[0]:
+                data = method(self, data)
+        elif target == "y":
+            for method in self.preprocessing_methods[1]:
+                data = method(self, data)
+        else:
+            raise ValueError("Target must be either 'X' or 'y'")
+        return data
+
+    def get_target_transformer(self) -> Pipeline:
         try:
-            return self._prior_predictive
+            return self.target_transformer  # type: ignore
         except AttributeError:
-            with self.model:
-                self._prior_predictive: az.InferenceData = pm.sample_prior_predictive(
-                    *args, **kwargs
-                )
-            return self._prior_predictive
+            identity_transformer = FunctionTransformer()
+            return Pipeline(steps=[("scaler", identity_transformer)])
 
-    def fit(
-        self,
-        progressbar: bool = True,
-        random_seed: RandomState = None,
-        *args: Any,
-        **kwargs: Any,
-    ) -> None:
-        with self.model:
-            self._fit_result = pm.sample(
-                progressbar=progressbar, random_seed=random_seed, *args, **kwargs
-            )
-            self._posterior_predictive = pm.sample_posterior_predictive(
-                trace=self._fit_result, progressbar=progressbar, random_seed=random_seed
-            )
+    @property
+    def prior_predictive(self) -> az.InferenceData:
+        if self.idata is None or "prior_predictive" not in self.idata:
+            raise RuntimeError("The model hasn't been fit yet, call .fit() first")
+        return self.idata["prior_predictive"]
 
     @property
-    def fit_result(self) -> az.InferenceData:
-        if self._fit_result is None:
+    # requesting discussion, is it something that we want to keep? Should I put on the deprecation warning to phase it out in later versions?
```

Comment:
Just for my understanding ... why should we remove it?

Reply:
honestly I didn't see much difference between calling baseobject.idata.posterior and baseobject.fit_result, so that's why I put the comment to make sure that it's something that we still want there.

Reply:
I don't have a strong opinion, but it makes more sense as a Github issue than as a comment in the code